File: 279.完全平方数.cpp
/*
 * @lc app=leetcode.cn id=279 lang=cpp
 *
 * dp[n]=k, 和为n的完全平方数的 最少数量k
 * dp[n] = 1 + min{ dp[n-1], dp[n-4], dp[n-9], ...}
 * [279] 完全平方数
 */

// @lc code=start
class Solution
{
public:
    int numSquares(int n)
    {
        vector<int> dp(n + 1, INT_MAX);
        dp[0] = 0;
        for (int i = 1; i <= n; ++i)
        {
            for (int j = 1; j * j <= i; ++j)
            {
                dp[i] = min(dp[i], dp[i - j * j] + 1);
            }
        }
        return dp[n];
    }
};
// @lc code=end

File: 279.完全平方数.py
#
# @lc app=leetcode.cn id=279 lang=python3
#
# dp[n] = 1 + min(dp[n-1],dp[n-4],...,dp[n-k*k])
#
# [279] 完全平方数
#

# @lc code=start
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [0] + [10000]*(n+1)
        for i in range(1, n+1):
            for k in range(1, int(math.sqrt(n))+1):
                dp[i] = min(dp[i], dp[i-k*k])
            dp[i] += 1
        return dp[n]
# @lc code=end
